bitcoin/src/network/constants.rs

```diff
@@ -97,18 +93,57 @@ impl Network {
     /// # Examples
     ///
     /// ```rust
-    /// use bitcoin::network::constants::Network;
+    /// use bitcoin::network::constants::{Network, Magic};
     ///
     /// let network = Network::Bitcoin;
-    /// assert_eq!(network.magic(), 0xD9B4BEF9);
+    /// assert_eq!(network.magic(), Magic::from_bytes([0xF9, 0xBE, 0xB4, 0xD9]));
     /// ```
-    pub fn magic(self) -> u32 {
-        // Note: any new entries here must be added to `from_magic` above
-        match self {
-            Network::Bitcoin => 0xD9B4BEF9,
-            Network::Testnet => 0x0709110B,
-            Network::Signet  => 0x40CF030A,
-            Network::Regtest => 0xDAB5BFFA,
+    pub fn magic(self) -> Magic {
+        Magic::from(self)
+    }
+}
+
+/// Network magic bytes to identify the cryptocurrency network the message was intended for.
+#[derive(Copy, PartialEq, Eq, PartialOrd, Ord, Clone, Hash)]
+pub struct Magic([u8; 4]);
+
+impl Magic {
+    /// Bitcoin mainnet network magic bytes.
+    pub const BITCOIN: Self = Self([0xF9, 0xBE, 0xB4, 0xD9]);
+    /// Bitcoin testnet network magic bytes.
+    pub const TESTNET: Self = Self([0x0B, 0x11, 0x09, 0x07]);
+    /// Bitcoin signet network magic bytes.
+    pub const SIGNET: Self = Self([0x0A, 0x03, 0xCF, 0x40]);
+    /// Bitcoin regtest network magic bytes.
+    pub const REGTEST: Self = Self([0xFA, 0xBF, 0xB5, 0xDA]);
+    
+    /// Create network magic from bytes.
+    pub fn from_bytes(bytes: [u8; 4]) -> Magic {
+        Magic(bytes)
+    }
+
+    /// Get network magic bytes.
+    pub fn to_bytes(self) -> [u8; 4] {
+        self.0
+    }
+}
+
+impl FromStr for Magic {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Magic, Self::Err> {
+        Ok(Magic(<[u8; 4]>::from_hex(s)?))
+    }
+}
+
+impl From<Network> for Magic {
+    fn from(network: Network) -> Magic {
+        match network {
+            // Note: Don't forget to add an entry to `try_from` below
```

Comment:
This comment is redundant now that `Magic` is an enum because the compiler will warn us if a new variant is added and not matched against.

Reply:
I believe the point is to warn us about `const` version which uses a trick instead of matching. Maybe the comment should says that more clearly.

Reply:
Maybe as @Kixunil suggested earlier, macros are the way to go.

Not an expert but would something like this be the way to go?
```rust
// network/constants.rs

// Note: Don't forget to add new networks and magic bytes below.
network_magic_conversion!(
    Network::Bitcoin => Magic::BITCOIN,
    Network::Testnet => Magic::TESTNET,
    Network::Signet => Magic::SIGNET,
    Network::Regtest => Magic::REGTEST
);

// internal_macros.rs

macro_rules! network_magic_conversion {
    (
        $($network:path => $magic:path),*
    ) => {
        impl From<Network> for Magic {
            fn from(network: Network) -> Magic {
                match network {
                    $($network => $magic,)*
                }
            }
        }

        impl TryFrom<Magic> for Network {
            type Error = UnknownMagic;

            fn try_from(magic: Magic) -> Result<Self, Self::Error> {
                match magic {
                    $($magic => Ok($network),)*
                    _ => Err(UnknownMagic(magic))
                }
            }
        }
    }
}
pub(crate) use network_magic_conversion;
```

Reply:
> @tcharding Magic is not an enum 

Oh boy, sorry man, I was up during the night and watching my kids while working this morning - face palm.

Reply:
At a glance the macro idea looks good but I'd say just leave it as is, the comments are fine - I'm just a wombat :)